Add draining control support to MergeHub #30057

nivox opened the pull request:
- MergeHub materialization now produces a DrainingControl object alongside the Sink
- The DrainingControl can be invoked to initiate the draining procedure
- No new producers can be connected to the Hub
- The Hub is completed as soon as all the currently connected producers complete

References #30057

nivox committed:
Add draining control support to MergeHub #30057

- MergeHub materialization now produces a DrainingControl object alongside the Sink
- The DrainingControl can be invoked to initiate the draining procedure
- No new producers can be connected to the Hub
- The Hub is completed as soon as all the currently connected producers complete

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

nivox commented:
@johanandren if you can take a look and give me some feedback I'd be grateful.

johanandren commented:
@nivox will probably not have time before next week, so please be patient :)

johanandren commented:
OK TO TEST

nivox commented:
@johanandren ok no problem

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -114,13 +170,20 @@ private[akka] class MergeHub[T](perProducerBufferSize: Int)
     private val queue = new AbstractNodeQueue[Event] {}
     @volatile private[this] var needWakeup = false
     @volatile private[this] var shuttingDown = false
+    private[this] var draining = false
```
Needs to be `@volatile` since it is used from the `Sink`s which are running on other threads than this stage.

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -237,20 +312,22 @@ private[akka] class MergeHub[T](perProducerBufferSize: Int)
+              } else {
+                completeStage()
+              }
+            } else completeStage()
```
make that a single `if (!isDraining && isShuttingDown)` instead?

nivox replied:
Agreed, I'll do that

johanandren reviewed (commented):
Left some feedback. 

Since the mergehub is used quite a bit in artery it would be good to verify it does not affect remoting throughput in a bad way. 

We don't have any microbench specifically for the MergeHub as far as I can see but there is a unit test with artery throughput tests that you can run with `akka-remote-tests/testOnly akka.remote.artery.MaxThroughputSpec`

johanandren commented:
Correction, running that test would be `akka-remote-tests/multi-jvm:testOnly akka.remote.artery.MaxThroughputSpec`

nivox commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -35,6 +33,21 @@ import akka.stream.stage._
+     * The returned [[Future]] will be completed as soon as the Hub finish processing all messages and completes.
+     * @return Hub completion future
+     */
+    def drainAndComplete(): Future[Done]
```
Yes, indeed the intent was more of having some kind of feedback that the draining was correctly initiated... 
I modeled that way following what has been done for the Alpakka Kafka draining support.

I'm not particularly invested in returning the completion future, what would be your suggestion for a different feedback mechanism?

nivox commented:
@johanandren  What about the fact that the CI tasks fails due to binary compatibility issues?

I thought keeping the public API stable and only changing the stage would be sufficient to maintain the binary compatibility.
Do you have any suggestions on that front?

I'll try and have a second pass on this PR during the week addressing the points you already mentioned

nivox committed:
Applying code review suggestions

nivox commented:
@johanandren I run the `MaxThroughputSpec` of the `master` branch against the PR and used the `PLOT_MaxThroughputSpec` lines to gather the results. The sample size is somewhat small as I only performed 11 runs on my laptop, in case let me know if I need to perform a more comprehensive test.

Here are the results:

- **warmup**:
  - *master*: **avg**=9.18087094998359 **median**=9.14795507344125
  - *PR*: **avg**=8.96331885273847 **median**=8.9379036214269

- **1-to-1**:
  - *master*: **avg**=23.8059643729237 **median**=23.9255975014112
  - *PR*: **avg**=24.3100631454149 **median**=24.5412845189462
  
- **1-to-1-size-1k**:
  - *master*: **avg**=205.90660139909 **median**=205.977174169816
  - *PR*: **avg**=206.420472876436 **median**=205.312016449139
 
- **1-to-1-size-10k**:
  - *master*: **avg**=452.378283884271 **median**=453.698533019148
  - *PR*: **avg**=440.229853787057 **median**=446.895181071345
  
- **5-to-5**:
  - *master*: **avg**=24.1057480773599 **median**=23.818039870286
  - *PR*: **avg**=23.2797982707817 **median**=23.2603491806402

Here's a plot of the whole dataset:
![throughput](https://user-images.githubusercontent.com/474133/110088308-fe4ca600-7d94-11eb-9062-d7035493f2ba.png)

My feeling is that the PR did not introduce any reduction in throughput. Let me know if you think I need to perform more tests.

lightbend-cla-validator commented:
At least one pull request committer is not linked to a user. See https://help.github.com/en/articles/why-are-my-commits-linked-to-the-wrong-user#commits-are-not-linked-to-any-user

johanandren commented:
Hi @nivox good work here, sorry for the slow reviewing, busy day. I'll hopefully have time to do a thorough review next week.

I think the bincomp changes can be filtered as there is no API a user can extend in there as far as I can see, you can read about how to filter those here: https://github.com/akka/akka/blob/master/CONTRIBUTING.md#binary-compatibility

The cla-validator message means that the email/user you have signed the CLA with is not the one you have in the commit messages, so please fix that (something like this: https://stackoverflow.com/questions/4981126/how-to-amend-several-commits-in-git-to-change-author/31669789#31669789 )

nivox committed:
Added MiMa binary compatibility exclusion

nivox committed:
Applied source formatting with scalafmt

johanandren reviewed (commented):
Added some comments, good work.

nivox committed:
Applied code review suggestions

nivox commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -114,13 +173,20 @@ private[akka] class MergeHub[T](perProducerBufferSize: Int)
         // We are only allowed to dequeue if we are not backpressured. See comment in tryProcessNext() for details.
         if (isAvailable(out)) tryProcessNext(firstAttempt = true))
 
+    private[MergeHub] val drainingCallback = getAsyncCallback[NotUsed] { _ =>
```
Making it conditional should actually be possible. I'll work on that.

nivox committed:
Adapted MergeHub tests to work without future feedback

nivox committed:
Made draining support AsyncCallback optional in MergeHub

nivox commented:
@johanandren I've changed the implementation to make the AsyncCallback used for draining support be created only when the functionality is required. This way the existing code relying on MergeHub shouldn't be impacted by the change at all.

Let me know what you think of this version and thanks for the support.

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -134,9 +200,16 @@ private[akka] class MergeHub[T](perProducerBufferSize: Int)
         true
       case Deregister(id) =>
         demands.remove(id)
+        if (draining) tryCompleteOnDraining()
```
`if (drainingEnabled && draining)` allows us to avoid touching the volatile when not enabled

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -157,10 +230,13 @@ private[akka] class MergeHub[T](perProducerBufferSize: Int)
         // and have been enqueued just after it
         if (firstAttempt)
           tryProcessNext(firstAttempt = false)
+        else if (draining)
```
Same here `if (drainingEnabled && draining)`

johanandren commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -157,10 +230,13 @@ private[akka] class MergeHub[T](perProducerBufferSize: Int)
     }
 
     def isShuttingDown: Boolean = shuttingDown
+    def isDraining: Boolean = draining
```
And here

johanandren reviewed (commented):
Looks like a good solution, I noted some additional optimizations based on that, avoiding touching the volatile if not supporting draining.

nivox committed:
Further optimizations in case MergeHub draining support is not enabled

nivox commented:
@johanandren I've integrated your optimizations regarding the access to the volatile `draining` variable when the draining support is not enabled.

johanandren reviewed (approved):
LGTM, thanks @nivox, great work!

jrudolph commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -49,7 +61,26 @@ object MergeHub {
+   * Completed or failed [[Sink]]s are simply removed. Once the [[Source]] is cancelled, the Hub is considered closed
+   * and any new producers using the [[Sink]] will be cancelled.
+   *
+   * The materialized [[DrainingControl]] can be used to drain the Hub: any new produces using the [[Sink]] will be cancelled
```
```suggestion
   * The materialized [[DrainingControl]] can be used to drain the Hub: any new producers using the [[Sink]] will be cancelled
```

jrudolph commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -64,14 +95,38 @@ object MergeHub {
+   * Completed or failed [[Sink]]s are simply removed. Once the [[Source]] is cancelled, the Hub is considered closed
+   * and any new producers using the [[Sink]] will be cancelled.
+   *
+   * The materialized [[DrainingControl]] can be used to drain the Hub: any new produces using the [[Sink]] will be cancelled
```
```suggestion
   * The materialized [[DrainingControl]] can be used to drain the Hub: any new producers using the [[Sink]] will be cancelled
```

jrudolph commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -35,6 +33,20 @@ import akka.stream.stage._
+   * A DrainingControl object is created during the materialization of a MergeHub and allows to initiate the draining
+   * and eventual completion of the Hub from the outside.
+   */
+  sealed trait DrainingControl {
```
Maybe it would make sense to call this trait `MergeHubControl` and also include the sink in it, so we can later add more functionality more easily without having to touch the API again.

nivox replied:
Indeed I tried to align the naming with Alpakka Kafka. As a user I think is good to have the API have similar names and behaviour as it lowers the cognitive burden. 

Regarding moving the `Sink` I second @octonato, I fear that it will make the API a little less obvious, again increasing the cognitive burden on the users.

johanandren replied:
I agree that keeping it aligned is good, so lets keep the name. I just noticed this one is missing the "not for user extension part" though. That allows us to add methods or change the hierarchy later, potentially introducing some other trait if needed. So please add that.

jrudolph reviewed (commented):
Some comments

octonato commented on akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala:
```diff
@@ -64,14 +95,38 @@ object MergeHub {
-private[akka] class MergeHub[T](perProducerBufferSize: Int)
-    extends GraphStageWithMaterializedValue[SourceShape[T], Sink[T, NotUsed]] {
+@InternalApi
+private[akka] final class MergeHubDrainingControlImpl(drainAction: () => Unit) extends MergeHub.DrainingControl {
```
if we keep the name `DrainingControl`, the impl can be called `DrainingControlImpl`. Like in the Java API.

octonato replied:
A common pattern we have is to have one impl class in an internal package that implements both Scala and Java interface. 

For example

https://github.com/akka/akka/blob/c5655a9ce605e3eb9982e8dd4d1e5a2c7d327e5e/akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala#L39-L45

But it seems that akka-stream is not following that pattern and the implementations are in `akka.stream.impl`.

Maybe the best here is to add the following in the `akka.stream.impl` package. 


```scala
private[akka] final class MergeHubDrainingControlImpl(drainAction: () => Unit) extends scaladsl.MergeHub.DrainingControl with javadsl.MergeHub.DrainingControl {
  override def drainAndComplete(): Unit = {
    drainAction()
  }
}
```

Note that in that case, it's better to preserve the name `MergeHubDrainingControlImpl` because it won't be under `MergeHub` anymore. 

We end up with one impl class that is good for both APIs and we don't need a delegate anymore.

octonato reviewed (approved):
Thanks @nivox, well done!

I left a few minor suggestions. 

There is an open discussion on how to name it (`MergeHubControl` or `MergeHub.DrainingControl`).

nivox committed:
Update akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

Co-authored-by: Johannes Rudolph <johannes.rudolph@gmail.com>

nivox committed:
Update akka-stream/src/main/scala/akka/stream/scaladsl/Hub.scala

Co-authored-by: Johannes Rudolph <johannes.rudolph@gmail.com>

nivox committed:
Update akka-stream/src/main/scala/akka/stream/javadsl/Hub.scala

Co-authored-by: Renato Cavalcanti <renato@cavalcanti.be>

nivox committed:
Update akka-stream/src/main/scala/akka/stream/javadsl/Hub.scala

Co-authored-by: Renato Cavalcanti <renato@cavalcanti.be>
